source_modelling/scripts/plot_mw_contributions.py

```diff
@@ -0,0 +1,98 @@
+        typer.Argument(help="Realisation filepath", dir_okay=False, exists=True),
+    ],
+    output_ffp: Annotated[
+        Path, typer.Argument(help="Output plot path.", writable=True, dir_okay=False)
+    ],
+    mu: Annotated[float, typer.Option(help="Shear rigidity constant")] = moment.MU,
+    dpi: Annotated[
+        float, typer.Option(help="Output plot DPI (higher is better).")
+    ] = 300,
+) -> None:
+    source_config: SourceConfig = realisations.read_config_from_realisation(
+        SourceConfig, realisation_ffp
+    )
+    rupture_propogation_config: RupturePropagationConfig = (
+        realisations.read_config_from_realisation(
+            RupturePropagationConfig, realisation_ffp
+        )
+    )
+    realisation_metadata: RealisationMetadata = (
+        realisations.read_config_from_realisation(RealisationMetadata, realisation_ffp)
+    )
+    total_area = sum(fault.area() for fault in source_config.source_geometries.values())
+    smallest_area = min(
+        fault.area() for fault in source_config.source_geometries.values()
+    )
+    area = np.linspace(smallest_area, total_area)
+    # Mw = log(area) + 3.995 is the Leonard2014 magnitude scaling relation
+    # for average rake.
+    plt.plot(
+        area, np.log10(area) + 3.995, label="Leonard 2014 Interplate (Average Rake)"
+    )
+
+    srf_data = srf.read_srf(srf_ffp)
+    total_magnitude = moment.moment_to_magnitude(
+        mu * (srf_data.points["area"] * srf_data.points["slip"] / (100**3)).sum()
+    )
+    plt.scatter(total_area, total_magnitude, label="Total Magnitude")
+
+    segment_counter = 0
+    point_counter = 0
+    for fault_name in rupture_propagation.tree_nodes_in_order(
+        rupture_propogation_config.rupture_causality_tree
+    ):
+        plane_count = len(source_config.source_geometries[fault_name].planes)
+        segments = srf_data.header.iloc[segment_counter : segment_counter + plane_count]
+
+        num_points = (segments["nstk"] * segments["ndip"]).sum()
+        individual_area = source_config.source_geometries[fault_name].area()
+
+        # get all points associated with all segments in the current fault
+        segment_points = srf_data.points.iloc[
+            point_counter : point_counter + num_points
+        ]
+        individual_magnitude = moment.moment_to_magnitude(
+            (mu * segment_points["area"] * segment_points["slip"] / (100**3)).sum()
+        )
+        plt.scatter(individual_area, individual_magnitude, label=fault_name)
+
+        # advance segment counter and point counter to skip all points from the current point
+        segment_counter += plane_count
+        point_counter += num_points
+
+    plt.xlabel("Area (m^2)")
+    plt.ylabel("Mw")
+    plt.xscale("log")
+    plt.legend()
+    plt.title(f"Log Area vs Magnitude ({realisation_metadata.name})")
+    plt.savefig(output_ffp, dpi=dpi)
+
+
+def main():
+    typer.run(plot_mw_contributions)
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
If you have just the 1 function in the script you can use the app and decorator and just call app instead of a new function main(), does allow for expansion for multiple functions if needed but unsure if it's "better" or not. When you do this you dont need to specify the function anyways as it's smart and just calls the only function in the file.

```python
app = typer.Typer()

@app.command(
    help="Some Function"
)
def some_function():
    pass

if __name__ == "__main__":
    app()```

Reply:
Yeah the distinction is actually for the `pyproject.toml` file so that I can create the `plot-*` scripts. It requires an entrypoint function that takes no arguments to call when the user types e.g. `plot-srf ...`. So I added the main function. The entrypoint only has to be `Callable`, however, so I could use the `app` variable instead. Maybe I will do this.